src/tests/shuttle.md

```diff
@@ -0,0 +1,205 @@
+## Testing concurrent code with Shuttle
+
+This note explains how to write concurrent tests for IPA using the Shuttle crate. IPA has been onboarded to Shuttle and
+there were already a couple of issues uncovered by using concurrency testing. Having more code tested by Shuttle reduces
+the chance of having intermittent failures in the future and makes software release more smooth.
+
+### Why does IPA need it?
+
+Testing concurrent code is hard because there is often an unbounded number of possible executions. There are multiple
+approaches to it: [exhaustive space exploration](https://docs.rs/loom/latest/loom/) and [stochastic simulation](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/asplos277-pct.pdf). Shuttle takes the latter approach and uses randomized
+scheduling techniques to detect concurrency issues with high probability. Certain IPA layers benefit more from having
+those tests
+
+#### Infrastructure
+Infrastructure uses tasks and schedulers to keep up with the load, so it benefits greatly from having good coverage
+of concurrency tests. In fact, [this issue](https://github.com/private-attribution/ipa/issues/256) was revealed only
+after onboarding infrastructure to Shuttle.
+
+Every module that uses `Arc`, `thread`, `tokio::spawn` benefits from concurrency testing.
+
+#### Protocols
+Protocols are inherently asynchronous and vulnerable to the same class of concurrency issues as the infrastructure.
+
+```rust
+async fn sum_of_products(a: &[u32], b: &[u32]) -> u64 {
+    let mut sum: u64 = 0;
+    for x in a {
+        for y in b {
+            let r = secure_mul(x, y).await;
+            sum += r;
+        }
+    }
+    
+    sum
+}
+```
```

Comment:
What does this example add to your message?

Reply:
yea I am puzzled too. Changed it to an example of concurrency bug